csharp/ql/src/semmle/code/csharp/PrintAst.qll

```diff
@@ -0,0 +1,606 @@
+}
+
+private predicate isInsideUnneededAttributable(Attributable attributable) {
+  isInsideUnneededType(attributable.(Field).getDeclaringType()) or
+  isInsideUnneededParameterizable(attributable.(Parameter).getDeclaringElement()) or
+  isInsideUnneededCallable(attributable.(Callable)) or
+  isInsideUnneededType(attributable.(DeclarationWithAccessors).getDeclaringType()) or
+  isInsideUnneededType(attributable.(ValueOrRefType))
+}
+
+private predicate isInsideUnneededUnboundGeneric(UnboundGeneric unboundGeneric) {
+  isInsideUnneededAttributable(unboundGeneric) or
+  isInsideUnneededParameterizable(unboundGeneric)
+}
+
+private predicate isInsideUnneededCallable(Callable callable) {
+  callable instanceof ConstructedGeneric or
+  isInsideUnneededType(callable.getDeclaringType())
+}
+
+private predicate isInsideUnneededType(Type type) {
+  type instanceof ConstructedType or
+  type.getDeclaringType*() instanceof ConstructedType or
+  type instanceof AnonymousClass or
+  type.getDeclaringType*() instanceof AnonymousClass
+}
+
+private predicate isInsideUnneededParameterizable(Parameterizable parameterizable) {
+  isInsideUnneededCallable(parameterizable) or
+  isInsideUnneededType(parameterizable.(Indexer).getDeclaringType()) or
+  isInsideUnneededType(parameterizable.(DelegateType))
+}
+
+private predicate isCompilerGeneratedParameterizable(Parameterizable parameterizable) {
+  parameterizable.isCompilerGenerated() or
+  parameterizable.getDeclaringType*().isCompilerGenerated()
+}
+
+private predicate isCompilerGeneratedAttributable(Attributable attributable) {
+  // Default parameter values on delegates are in the tree multiple times
+  // due to the compiler generated `Invoke`.
+  isCompilerGeneratedParameterizable(attributable.(Parameter).getDeclaringElement())
+}
+
+/**
+ * Retrieves the canonical QL class(es) for entity `el`
+ */
+private string getQlClass(Element el) {
+  result = "[" + concat(el.getAPrimaryQlClass(), ",") + "] "
+  // Alternative implementation -- do not delete. It is useful for QL class discovery.
+  // result = "["+ concat(el.getAQlClass(), ",") + "] "
+}
+
+/**
+ * An `Element`, such as a `namespace` and a `partial class`, might have multiple locations.
+ * The locations are ordered by line, column, and then the first one is selected.
+ */
+private Location getRepresentativeLocation(Element ast) {
+  result =
+    rank[1](Location loc |
+      loc = ast.getLocation() and selectedFile(loc.getFile())
+    |
+      loc order by loc.getStartLine(), loc.getStartColumn(), loc.getEndLine(), loc.getEndColumn()
+    )
+}
+
+private predicate locationSortKeys(Element ast, string file, int line, int column) {
+  if exists(getRepresentativeLocation(ast))
+  then
+    exists(Location loc |
+      loc = getRepresentativeLocation(ast) and
+      file = loc.getFile().toString() and
+      line = loc.getStartLine() and
+      column = loc.getStartColumn()
+    )
+  else (
+    file = "" and
+    line = 0 and
+    column = 0
+  )
+}
```

Comment:
This can be simplified to
```ql
  exists(Location loc |
    loc = getRepresentativeLocation(ast) and
    file = loc.getFile().toString() and
    line = loc.getStartLine() and
    column = loc.getStartColumn()
  )
  or
  not exists(getRepresentativeLocation(ast)) and
  file = "" and
  line = 0 and
  column = 0
```